### PR TITLE
ARROW-1839/ARROW-1871: [C++/Python] Add Decimal Parquet Read/Write Tests

### PR DESCRIPTION
**Pull request by cpcloud**

### CODE REVIEW

**Comment by wesm** [comment]
Rebased

**Comment by cpcloud** [comment]
This isn't quite ready to go I have a few more things to add.

**Comment by cpcloud** [comment]
@wesm This is no longer a WIP and is ready for review.

I have a test (`test_decimal_roundtrip_negative_scale`) that depends on https://issues.apache.org/jira/browse/PARQUET-1167. I can remove that here and add in a later PR if necessary.

File: python/pyarrow/tests/test_parquet.py
Comment (cpcloud) [inline python/pyarrow/tests/test_parquet.py]: This is generating a ton of tests, I will move the `scale` inside the function.

**Comment by wesm** [comment]
I force-pushed this after PARQUET-1167 was merged so we can get a passing build

File: cpp/src/arrow/python/builtin_convert.cc
Comment (wesm) [inline cpp/src/arrow/python/builtin_convert.cc]: I'm not sure why these were ever virtual. Making them non-virtual because we're already using CRTP here

File: cpp/src/arrow/util/decimal.h
Comment (wesm) [inline cpp/src/arrow/util/decimal.h]: I'm renaming this to `RescaleDecimal` since it might be ambiguous in the `arrow` namespace otherwise

**Review by wesm — APPROVED** [review]
+1, OK to merge once the build is passing

File: cpp/src/arrow/python/builtin_convert.cc
Comment (cpcloud) [reply cpp/src/arrow/python/builtin_convert.cc]: Thanks, sounds good to me

**Comment by wesm** [comment]
There's an odd Python 2.7 failure here

**Comment by cpcloud** [comment]
Yep I know what it is. Putting up a fix in a bit. Pylong_check is for actual long typed values. Py3 only has int checking

**Comment by wesm** [comment]
@cpcloud `RescaleDecimal` wasn't exported, so instead I made it an instance method of `Decimal128` if that's cool with you

**Comment by cpcloud** [comment]
@wesm Yep that's fine with me. We will need this for decimal kernels pretty soon.

**Comment by wesm** [comment]
Unrelated Travis CI failure. I'll merge as soon as we have a green Appveyor build

**Comment by wesm** [comment]
Appveyor finally passing: https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/build/1.0.4623. Merging! thanks @cpcloud for your herculean efforts on this!

**Pull request closed without merging**
